부산대 BE_오선재 - Step3 위시 리스트

5seonjae opened the pull request:
# PR — “Step 3: Wish 기능 구현 (API + 최소 UI + 테스트 완비)”

## ✨ 구현 범위

| Layer | 주요 클래스 / 템플릿 | 핵심 기능 |
|-------|---------------------|-----------|
| **Auth** | `@LoginMember`, `LoginMemberArgumentResolver`, `BearerAuthHeaderParser`, `WebConfig` | • Bearer JWT 추출&검증<br>• 컨트롤러 파라미터로 `Member` 주입 |
| **Domain** | `WishItem` (POJO) | 불변 도메인 객체 (memberId, productId, …) |
| **Repository** | `WishRepository` | • `findAllByMemberId`<br>• `updateOrInsertWishItem` (수량 + 1)<br>• `delete` |
| **Service** | `WishService` | • `getWishListForMember`<br>• `addWishItemForMember` (존재 확인 포함)<br>• `removeWishItemForMember` |
| **REST API** | `WishController` (`/api/wishes`) | GET · POST · DELETE 엔드포인트 |
| **View (UI)** | `WishViewController` (`/wishes`)<br>`templates/wishes/list.html` | • 목록 테이블·삭제 버튼<br>• 상품 페이지에 🤍 폼 추가 (`action="/wishes/{id}"`) |
| **DTO** | `WishRequestDto`, `WishResponseDto` | 요청 / 응답 전용 레코드 (@Valid 포함) |
| **테스트** | `WishControllerTest` (API E2E 14 case)<br>`WishViewControllerTest` (View E2E 10 case) | 인증/예외/모델 검증 전부 **GREEN** |

## 📂 패키지 구조(요약)
```text
src/main/java/gift
├─ auth
│  ├─ LoginMember.java
│  └─ LoginMemberArgumentResolver.java
├─ config
│  └─ WebConfig.java
├─ controller
│  ├─ WishController.java
│  └─ WishViewController.java
├─ dto/api
│  ├─ WishRequest.java
│  └─ WishItemResponse.java
├─ entity           └─ WishItem.java
├─ repository       ├─ WishRepository.java
│                   └─ ProductRepository.java
├─ service          └─ WishService.java
└─ util             └─ BearerAuthHeaderParser.java

src/main/resources/templates
├─ products/list.html   ← 🤍 폼
└─ wishes/list.html
```

---

## 🧪 테스트 결과

API E2E : 헤더 누락·프리픽스 오류·검증 실패·정상 흐름 전부 커버

View E2E : 모델 데이터 매칭(model().attribute(..., contains(hasProperty(...)))) 및 리다이렉트 검증

## 🔍 검증 가이드
1. API ( 예 : Postman )

- 로그인 (Step 1 API)
POST /api/members/login          Authorization: Basic base64(id:pw)
→ { "token": "eyJ…" }

- 위시 목록
GET /api/wishes                  Authorization: Bearer <token>

- 담기
POST /api/wishes
Authorization: Bearer <token>
Content-Type: application/json
{ "productId": 3, "quantity": 1 }

- 삭제
DELETE /api/wishes/3             Authorization: Bearer <token>

## ⚠️ 남은 과제 & 일정

| TODO | 설명 |
| -- | --  |
| 로그인/회원가입 화면 | Form → AUTH 쿠키 발행 |
| Admin·User 네비 분기 | member.admin 조건 렌더링 |
| 메인(/home) | 로그인 전용 랜딩 + 메뉴 |

시간 관계상 위 3개 화면은 이번 PR에 포함하지 못했습니다.
기능 로드맵과 ETA를 위와 같이 명시했으며, 차기 PR에서 처리하겠습니다.

5seonjae committed:
docs: README에 위시 리스트 기능 스펙 및 체크리스트 추가

- Step 3 진행 체크리스트 추가 (`wish_items` 테이블 마이그레이션, Repository/Service/Controller, 테스트, UI 등)
- `wish_items` 테이블 스키마 및 마이그레이션 SQL 예시 삽입
- 위시 리스트 조회·추가·삭제 API 명세(메서드, 경로, 요청/응답 포맷, 상태 코드, 예시) 업데이트

5seonjae committed:
feat: wish_items 테이블 생성

- `id`, `member_id`, `product_id`, `quantity` 컬럼 정의
- `member_id`와 `product_id` 조합에 고유 제약(`uk_member_product`) 적용
- `members(id)`, `products(id)` 외래 키 참조 추가

5seonjae committed:
feat: WishRequestDto record 생성 및 유효성 검증 추가

- `gift.dto.api.WishRequestDto` record 클래스 생성
- `productId` 필드에 `@NotNull(message = "상품 ID는 필수입니다.")` 검증 추가
- `quantity` 필드에 `@NotNull(message = "상품 수량은 필수입니다.")` 및 `@Min(value = 1, message = "한 개 이상 담을 수 있습니다.")` 검증 추가

5seonjae committed:
feat: 위시 리스트 API 응답용 WishResponseDto 레코드 추가

- 상품 ID(productId), 이름(name), 가격(price), 이미지 URL(imageUrl), 수량(quantity) 필드를 갖는 WishResponseDto 레코드 클래스 생성
- Java Record 문법을 활용해 불변(immutable) DTO로 구현하여 간결성을 확보
- gift.dto.api 패키지에 위치시켜 API 계층의 응답 모델과 명확히 분리
- 위시 리스트 조회 엔드포인트(/api/wishes 등)에서 클라이언트에 전달할 데이터 구조 정의

5seonjae committed:
feat(auth): 로그인 사용자 자동 바인딩용 @LoginMember 애노테이션 추가

- gift.auth 패키지에 @LoginMember 애노테이션 정의
- @Target(ElementType.PARAMETER), @Retention(RetentionPolicy.RUNTIME) 설정으로 컨트롤러 메서드 파라미터에 적용 가능
- Argument Resolver 구현 시, 해당 어노테이션을 통해 로그인된 멤버 객체를 간편하게 주입할 수 있는 기반 마련
- LoginMemberArgumentResolver, WebConfig, BearerAuthHeaderParser, WishController, WishRepository, WishService 빈 클래스 플레이스홀더 생성

5seonjae committed:
feat(config): LoginMemberArgumentResolver 등록을 위한 WebConfig 추가

- gift.config.WebConfig 클래스 생성
- WebMvcConfigurer 구현 및 addArgumentResolvers 오버라이드하여 LoginMemberArgumentResolver를 argumentResolvers 리스트에 등록
- @LoginMember 애노테이션이 적용된 컨트롤러 파라미터에서 자동으로 리졸버가 동작하도록 설정

5seonjae committed:
feat(auth): @LoginMember 지원을 위한 ArgumentResolver 및 TokenService JWT 파싱 로직 추가

- LoginMemberArgumentResolver 클래스 추가
    - supportsParameter(): @LoginMember 애노테이션과 Member 타입 파라미터 처리
    - resolveArgument():
        1. HTTP 요청 헤더 ‘Authorization’에서 Bearer 토큰 추출
        2. TokenService.parseClaims() 호출해 JWT Claims 파싱
        3. Claims.getSubject()로 얻은 memberId로 MemberRepository.findById 조회
        4. 회원 미존재 시 IllegalStateException 발생
- TokenService에 parseClaims(String token) 메서드 추가
    - Jwts.parser() 기반 JWT 파싱 및 exp 검증 로직 구현
- MemberRepository에 findById(Long id) 메서드 추가 (JdbcClient 활용 SQL 쿼리 및 ResultSet 매핑)
- BearerAuthHeaderParser 컴포넌트 추가
    - 헤더 누락 시 MissingAuthorizationHeaderException, 형식 오류 시 InvalidAuthorizationHeaderException 발생

5seonjae committed:
chore: 플레이스홀더 빈 클래스 삭제

5seonjae committed:
feat(wish): JdbcClient 기반 WishRepository 초기 구현

- gift.repository 패키지에 WishRepository 클래스 생성 및 @Repository 애노테이션 적용
- JdbcClient를 주입받는 생성자 구현
- CRUD 메서드 골격만 정의, 실제 구현은 추후 작업 예정

5seonjae committed:
feat(wish): WishService 클래스 생성 및 DI 스켈레톤 추가

- gift.service 패키지에 WishService 클래스 생성 후 @Service 애노테이션 적용
- 생성자 주입 방식으로 WishRepository 의존성 주입 구현
- 향후 위시 비즈니스 로직 구현을 위한 기본 골격 마련

5seonjae committed:
feat(wish): WishController REST API 엔드포인트 스켈레톤 추가

- gift.controller 패키지에 WishController 클래스 생성 및 @RestController 적용
- 클래스 레벨에 @RequestMapping("/api/wishes") 설정으로 기본 URL 경로 지정
- 생성자 주입 방식으로 WishService 의존성 주입 구현

5seonjae committed:
feat(wish): WishItem 도메인 엔티티 클래스 추가

- gift.entity 패키지에 WishItem 클래스 생성
- memberId, productId, name, price, imageUrl, quantity 필드를 final로 선언해 불변 객체 구현
- 모든 필드를 파라미터로 받는 생성자 및 getter 메서드만 포함

5seonjae committed:
fix(schema): wish_items 테이블 제약 조건 구문 오류 수정

- CONSTRAINT 선언부의 중복·위치 오류를 수정하여 DDL 실행 시 문법 에러 해결
    - UNIQUE 제약(uk_member_product) 정의
    - FOREIGN KEY 제약(fk_wish_member, fk_wish_product) 정의
- AUTO_INCREMENT, NOT NULL, DEFAULT 1 등 기존 스펙은 유지

5seonjae committed:
test(wish): Wish 목록 API 통합 테스트 및 실제 조회 로직 구현

- **WishControllerTest**
    - MockMvc 기반 통합 테스트 5종 작성
        1. 정상 조회 200 OK + JSON 배열
        2. 빈 목록 조회
        3. 복수 아이템 조회
        4. Authorization 누락 → 401
        5. 잘못된 프리픽스 → 401
    - `@MockitoBean` 으로 WishService·LoginMemberArgumentResolver 목킹

- **WishService**
    - `getWishListForMember(Member)` 메서드 구현
    - WishItem → WishResponseDto 매핑 전용 메서드 추가

- **WishRepository**
    - `findAllByMemberId(Long)` 구현
        (wish_items ↔ products 조인하여 WishItem 목록 반환)

- **WishController**
    - `GET /api/wishes` 엔드포인트에서 WishService 호출해 DTO 리스트 반환

5seonjae committed:
chore(db): H2 URL을 MySQL 호환 모드로 변경

- spring.datasource.url
    jdbc:h2:mem:test
        → jdbc:h2:mem:testdb;MODE=MySQL;DB_CLOSE_DELAY=-1
- MODE=MySQL 로 MySQL 구문·데이터타입 호환성 확보
- DB_CLOSE_DELAY=-1 로 테스트 실행 중 메모리 DB 유지

5seonjae committed:
test(wish): WishController 통합 테스트로 전환 및 커버리지 향상

- @SpringBootTest + @AutoConfigureMockMvc 조합으로 통합 테스트 구성
    - 실제 Bean(WishService, MemberRepository, ProductRepository) 주입
    - @Transactional 적용으로 메모리 DB 롤백 보장
- LoginMemberArgumentResolver만 @MockitoBean 으로 목킹해 인증 흐름 제어
- 통합 시나리오 5종 구현
    1. 정상 조회(200 OK + 목록 JSON)      :contentReference[oaicite:0]{index=0}
    2. 빈 목록 조회                       :contentReference[oaicite:1]{index=1}
    3. 다건 조회                          :contentReference[oaicite:2]{index=2}
    4. Authorization 헤더 누락 → 401      :contentReference[oaicite:3]{index=3}
    5. 잘못된 프리픽스(Basic) → 401      :contentReference[oaicite:4]{index=4}
- MockMvc 활용으로 컨트롤러·서비스·리포지토리 연동 및 JSON 응답 검증

5seonjae committed:
fix(wish): WishRepository 조회 SQL 문법 오류 수정

- SELECT 절 문자열에서 잘못된 개행·세미콜론 문제 해결
- H2(MySQL 모드)에서도 구문 오류 없이 실행되도록 수정
- findAllByMemberId 통합 테스트 정상 통과 확인

5seonjae committed:
feat(wish): 위시 리스트 상품 추가 API 구현 및 통합 테스트 작성

- **WishController**
    - `POST /api/wishes` 엔드포인트 구현 (`201 Created`)
    - `@LoginMember` 인증 멤버 주입 + `@Valid WishRequestDto` 바인딩
- **WishService**
    - `addWishItemForMember()` 추가
        - 상품 존재 여부 확인, 없으면 `NoSuchElementException` → 404
        - WishRepository에 upsert 호출
- **WishRepository**
    - `updateOrInsertWishItem(memberId, productId, quantity)`
        - `INSERT … ON DUPLICATE KEY UPDATE` 로 수량 누적

- **WishControllerTest** 통합 테스트 7종
    1. productId 누락/빈 값 → 400
    2. quantity 누락/빈 값/0 이하 → 400
    3. 존재하지 않는 상품 ID → 404
    4. 인증 헤더 누락 → 401
    5. 잘못된 Authorization 프리픽스(Basic) → 401
    6. JSON 형식 오류 → 400
    7. 정상 추가 시 201 Created

5seonjae committed:
feat(wish): 위시 리스트 상품 삭제 API 구현 및 통합 테스트 추가

- **WishController**
    - `DELETE /api/wishes/{productId}` 엔드포인트 구현
    - 인증된 멤버(@LoginMember)와 PathVariable productId로 삭제 수행
    - 성공 시 204 No Content 반환
- **WishService**
    - `removeWishItemForMember(member, productId)` 추가
- **WishRepository**
    - `DELETE FROM wish_items WHERE member_id = ? AND product_id = ?` 실행하는 `delete()` 메서드 구현

- **WishControllerTest**
    1. 정상 삭제 → 204 No Content
    2. 존재하지 않는 상품 ID → 404 Not Found
    3. Authorization 헤더 누락 → 401 Unauthorized
    4. 잘못된 프리픽스(Basic) → 401 Unauthorized

5seonjae committed:
fix(wish): 위시 상품 삭제 로직 보강 및 Repository delete 결과값 반환

WishService
- delete 실행 결과가 0행이면 `NoSuchElementException("위시 목록에 없는 상품입니다.")` 발생하도록 검증 추가

WishRepository
- `delete(memberId, productId)`가 `jdbcClient.update()` 결과(영향된 행 수)를 그대로 반환하도록 구현

5seonjae committed:
feat(view): 위시 리스트 MVC 뷰 레이어 완성 및 통합 테스트 추가

기능 추가
- **WishViewController**
    - `/wishes` 목록 페이지, 상품 추가 폼(`POST /wishes/{productId}`), 삭제 폼(`POST /wishes/{productId}/delete`) 구현
    - `@LoginMember` 인증·`RedirectAttributes` 플래시 메시지 처리

- **Thymeleaf 템플릿**
    - `templates/products/list.html`에 🤍 찜 버튼 추가
    - `templates/wishes/list.html` 신규 작성 ― 목록·빈 상태·삭제 버튼 UI 포함

- **WishResponseDto**
    - Thymeleaf 바인딩용 getter 5종 추가

테스트
- **WishViewControllerTest**
    - `@SpringBootTest + MockMvc + @Transactional`
    - 로그인 목킹용 `LoginMemberArgumentResolver`를 `@MockitoBean`으로 주입
    - 시나리오 9종 검사
        1. 위시 추가 → 302 Found
        2. 헤더 누락 → 401
        3. 잘못된 프리픽스 → 401
        4. 존재하지 않는 상품 → 404
        5. 목록 1 건 → 모델 검증
        6. 목록 2 건 → 모델 검증
        7. 목록 요청 헤더 누락 → 401
        8. 정상 삭제 → 302 Found
        9. 삭제 대상 미존재 → 404

기타
- README 위시 기능 체크리스트 완료 항목 업데이트

5seonjae committed:
chore(asset): 초기 상품 이미지 외부 의존 제거 및 정적 리소스 추가

- AI로 생성한 상품 이미지 2종을 `resources/static/image/` 폴더에 추가
    · chocolate.webp
    · shrimpCrackers.webp
- `data.sql` 초기 데이터의 `image_url` 값을
    `http://localhost:8080/image/{filename}.webp` 형식으로 교체
    → 외부 URL 주기적 갱신으로 발생하던 이미지 깨짐 문제 해결

5seonjae committed:
feat(view+auth): 로그인/회원가입 MVC 뷰 추가 및 권한별 상품 목록 분리

- **gift.controller.api**
    MemberController, ProductController, WishController 이동
- **gift.controller.view**
    - AuthViewController (신규)
    - ProductViewController – 사용자 전용 (신규)
    - AdminProductViewController – 기존 관리자용 컨트롤러 이름·경로 변경
    - WishViewController 경로 이동

- `members` 테이블에 `is_admin BOOLEAN NOT NULL DEFAULT FALSE` 컬럼 추가
- `data.sql`
    - 관리자 계정 `5seonjae@gmail.com`
    - 일반 사용자 `6seonjae@gmail.com`  등록

- **AuthViewController**
    - `/signup`, `/login`, `/logout` 뷰·로직 구현
    - 로그인 성공 시 `jwt` 발급 후
        - 관리자는 `/admin/products`
        - 일반 사용자는 `/products` 리다이렉트
- **MemberRepository/WishService**
    - `isAdminMember()` 구현해 관리자 여부 판별

- `auth/login.html`, `auth/signup.html` 신규
- 관리자용 템플릿 `products/admin/*`  디렉터리로 이동
- 사용자용 템플릿 `products/user/detail.html`, `products/user/list.html` 신규
- `main.css` 버튼 글자색 `#fff → black`

- **AdminProductViewControllerTest** (이전 ProductViewControllerTest 이름 변경)
- **ProductViewControllerTest** – 사용자 전용 통합 테스트 신규
    - 목록·상세·예외(404) 시나리오 검증

- WishRepository `isAdminMember` SQL 추가
- ProductRepository 테스트 초기화 코드 보강

BREAKING CHANGE: 컨트롤러 패키지 경로가 변경(gift.controller.api/view)되어
기존 import 경로와 URL 매핑이 달라졌습니다.

5seonjae commented:
[feat(view+auth): 로그인/회원가입 MVC 뷰 추가 및 권한별 상품 목록 분리](https://github.com/next-step/spring-gift-wishlist/pull/287/commits/e034bcd09a7fa83e01d06b2eb57f95b07e47b69e)

이 커밋에 변경 파일 수가 많아서 리뷰가 불편하실 수 있다는 점 먼저 양해 부탁드립니다. 🙇🏻‍♂️

1. 커밋이 커진 배경

| 이유 | 상세 설명 |
| -- | -- |
| 패키지/경로 리팩터링 ↔ 기능 작업의 동시 진행 | 로그인·회원가입 뷰 추가와 동시에 API · 뷰 컨트롤러를 controller.api / controller.view 로 분리하느라 파일 이동·import 수정이 대거 발생했습니다. |
| 권한 흐름(관리자/사용자) 분기 | URL·템플릿 경로가 한꺼번에 바뀌면서 기존 테스트·서비스 코드도 전방위 수정이 필요했습니다. |
| 타임 프레임 제약 | 단계별 커밋을 나누는 동안 빌드가 계속 깨져 리뷰 블로킹이 길어질 위험이 있어, ‘전체 기능 완성 + 빌드 통과’를 우선시하여 한 커밋으로 묶었습니다. |

2. 변경 내역 요약

| 카테고리 | 핵심 변경 |
| -- | -- |
| AuthView | /signup, /login, /logout 뷰·컨트롤러 및 JWT 발급 플로우 |
| 권한 분기 | 관리자는 /admin/products, 일반 사용자는 /products 로 리다이렉트 |
| 패키지/경로 리팩터링 | gift.controller.api/*, gift.controller.view/* 로 이동 |
| 뷰 템플릿 | 관리자인지에 따라 분리된 목록·상세 페이지, 공통 레이아웃 수정 |
| 스키마 & seed | members.is_admin 컬럼 추가, 관리자/사용자 초기 데이터 |
| 테스트 | Admin / User 전용 컨트롤러 테스트 분리, 권한 시나리오 보강 |

리뷰에 불편을 드려 죄송합니다. 🙏
혹시 보시다 막히는 부분이나 추가적으로 쪼개서 올렸으면 하는 파트가 있으면 코멘트로 알려 주세요. 다음 PR부터는 커밋 단위를 작게 유지하도록 하겠습니다!

hjjae2 commented on src/main/resources/application.properties:
```diff
@@ -3,7 +3,7 @@ spring.config.import=file:.env[.properties]
 
 # H2 DB 설정
-spring.datasource.url=jdbc:h2:mem:test
+spring.datasource.url=jdbc:h2:mem:testdb;MODE=MySQL;DB_CLOSE_DELAY=-1
```
`DB_CLOSE_DELAY=-1` 설정은 왜 추가가 되었나요~?

5seonjae replied:
### DB_CLOSE_DELAY=-1을 추가한 이유

1. “마지막 커넥션이 닫히면 DB 자체가 언로드된다”는 H2 기본 동작을 막기 위해서입니다.

- H2 는 활성 세션 수가 0 이 되는 순간 인-메모리 DB 를 바로 메모리에서 제거합니다.
- Spring Boot 기동 과정 ( 스키마 생성 후 풀 정리, @SpringBootTest 컨텍스트 재사용, DevTools Hot Restart 등 ) 에서 잠시라도 커넥션이 모두 닫히면 테이블을 찾지 못하는 오류가 간헐적으로 발생할 수 있습니다.
- `;DB_CLOSE_DELAY=-1` 을 붙이면 세션이 0 이어도 JVM 이 살아 있는 동안 DB 를 보존하도록 설정됩니다. 

2. 메모리 누수 우려는 개발·테스트 범위에서 거의 없다

- H2 공식문서에 따르면, “may create a memory leak” 경고가 있지만, 스키마·데이터가 크지 않고애플리케이션(JVM) 종료 시 메모리가 전부 반납되므로 실제 문제가 될 가능성은 낮습니다.
- 혹시 장시간 실행하며 대용량 데이터를 쌓을 경우에는 종료 훅 ( @PreDestroy ) 에서 SHUTDOWN 명령을 호출하거나 CI/Prod 프로필에서는 DB_CLOSE_DELAY=0 ( 기본값 ) 으로 돌려 메모리를 즉시 회수할 수 있습니다.
  - `@PreDestroy` : Spring Bean 이 소멸 직전 ( 컨텍스트가 내려가거나 애플리케이션이 종료될 때 ) 호출되는 라이프사이클 콜백
  - `SHUTDOWN` : H2 전용 SQL 명령 / 실행하면 해당 DB 를 명시적으로 닫고, 메모리·파일 핸들을 즉시 해제

3. 결론

- 개발·통합 테스트 안정성을 높이기 위한 안전장치이며, 운영에는 적용되지 않아 성능·메모리 이슈가 없습니다.
- 필요 시 프로필 분리로 더 세밀하게 제어하거나, SHUTDOWN 명령으로 수동 정리할 수 있습니다.

hjjae2 commented on src/test/java/gift/WishControllerTest.java:
크, 테스트 코드를 꼼꼼하게 잘 작성해주셨네요. 든든합니다. 👍

hjjae2 commented on src/main/java/gift/auth/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,59 @@
+import org.springframework.web.method.support.ModelAndViewContainer;
+
+@Component
+public class LoginMemberArgumentResolver
+    implements HandlerMethodArgumentResolver {
```
HandlerMethodArgumentResolver 를 잘 활용하셨군요. 👍

hjjae2 commented on src/main/java/gift/auth/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,59 @@
+                                  NativeWebRequest webRequest,
+                                  WebDataBinderFactory binderFactory) {
+
+        // Authorization 헤더 추출
+        String header = webRequest.getHeader("Authorization");
+
+        // Authorization 헤더에서 Bearer 스킴 검증 후 토큰 문자열만 추출
+        String token = authHeaderParser.extractBearerToken(header);
+
+        // JWT 파싱 및 Claim 추출
+        Claims claims = tokenService.parseClaims(token);
```
주석 없이도 코드 파악에 차질이 없는데요, 주석의 작성 이유가 무엇일까요?

5seonjae replied:
### 주석을 달았던 이유

제가 이 부분에 주석을 달았던 이유는 코드에 대해 개인적인 의도를 기록하고 싶었기 때문이었습니다. 처음엔 “나중에 이 코드를 봤을 때 Bearer  스킴 → 토큰 → Claim 추출 순서를 바로 떠올릴 수 있도록” 메모 차원에서 달았습니다.

### 하지만 제거 ( 또는 위치 이동 ) 가 더 좋다고 판단한 이유

메서드 네이밍이 충분히 구체적이고, 로직이 바뀔 때 주석 동기화 필요하기 때문에, 유지보수에 부담이 될 수 있습니다. 

### 정리

이번 PR에선 주석을 삭제하고, 추후에 맥락이 필요한 부분에 대해서는 Javadoc·PR 설명으로 옮기겠습니다.
이후에도 “코드로 충분히 드러나면 주석 대신 리팩터링” 원칙을 더 신경 쓰겠습니다!

hjjae2 replied:
> 이후에도 “코드로 충분히 드러나면 주석 대신 리팩터링” 원칙을 더 신경 쓰겠습니다!

제가 전달하고 싶은 메시지이기도 했는데요, 좋네요! 👍

hjjae2 commented on src/main/java/gift/controller/api/WishController.java:
```diff
@@ -0,0 +1,43 @@
+        wishService.addWishItemForMember(member, wishRequestDto);
+    }
+
+    @DeleteMapping("/{productId}")
```
REST API 에선, 경로에서 리소스(= 표현)간의 계층 구조/관계도를 표현할 수 있어 URL 만 보더라도 대략적인 요청과 응답을 예측할 수 있습니다.

`/api/wishes/1` 라는 경로가 있다면, 1은 바로 앞에 위치한 `wishes` 에 종속된, 관계가 있는 정보라고 이해할 수 있습니다.

그런데 설계해주신 건 `wishes` 와 관계가 없어보이는(`product` 와 관련이 있는) 정보가 오도록 설계해주셨는데요.

URL 만 보고 `product`의 식별자, `member`의 식별자인지 클라이언트가 식별할 수 있을까요?

5seonjae replied:
피드백 감사드립니다! URL 계층 구조를 놓친 부분 인정합니다 🙏

REST URL 은 ‘슬래시(/) = 계층(hierarchy)’ 라는 암묵적 규칙이 있습니다.
예를 들면,
```Plain Text
/posts/{postId}/comments/{commentId}
```
만 봐도 comment 가 post 에 종속된다는 것을 알 수 있습니다.

제가 작성했던
```Plain Text
/api/wishes/{productId}
```
는 productId 가 실제로는 “특정 member  의 Wishlist 안에 있는 상품 ID” 라는 맥락을 드러내지 못한다는 점을 간과했습니다.

---

이 부분을 참고하여 이런 식으로 수정하려고 합니다.

- 기존: `POST /api/wishes/{productId}`, `DELETE /api/wishes/{productId}`
- 변경: `POST /api/wishes/product/{productId}`, `DELETE /api/wishes/product/{productId}`

`product/` 세그먼트를 추가해 마지막 식별자가 상품 ID 라는 점을 URL 로 명확히 표현하겠습니다.

hjjae2 commented on src/main/java/gift/entity/WishItem.java:
```diff
@@ -0,0 +1,50 @@
+
+    private final Long memberId;
+    private final Long productId;
+    private final String name;
```
사소한데요, 어떤 이름인지 헷갈릴 수 있지는 않을까요-?

(현재 `Member`가 `name` 이라는 속성을 갖고 있진 않지만, 통상 사용될 수 있으므로) 
`Member` 의 이름인지, `Product` 의 이름인지, 혹은 또 다른 개념의 이름인지 와 같이요.

5seonjae replied:
다시 보니 name 필드가 “무엇의 이름인지” 모호하다는 지적이 맞습니다.
저 의도는 “해당 시점의 상품명 스냅샷” 이었는데, 코드만 봐서는 Member 이름 / Product 이름 / 기타 라벨로 모두 해석될 여지가 있습니다.

Product 엔티티의 필드를 그대로 가져와서 사용하려 하다보니 이 부분에 대해 미처 생각을 못했던 것 같습니다...

---

필드명을 name 에서 productName 으로 변경하도록 하겠습니다.

필드명·컬럼명 수정 후 다시 푸시하겠습니다. 추가 의견 있으시면 알려주세요!

hjjae2 commented on src/main/java/gift/service/WishService.java:
```diff
@@ -0,0 +1,67 @@
+        this.productRepository = productRepository;
+    }
+
+    private WishResponseDto mapWishItemToWishResponseDto(WishItem wi) {
+        return new WishResponseDto(
+                wi.getProductId(),
+                wi.getName(),
+                wi.getPrice(),
+                wi.getImageUrl(),
+                wi.getQuantity()
+        );
+    }
```
`WishResponseDto` 의 정적 팩터리 메서드로 만들어보면 어때요?

```java
static WishResponseDto of(WishItem wi) {
    ...
}
```

5seonjae replied:
피드백 감사드립니다! 내부적으로도 “DTO 자체에서 변환 책임을 갖게 할까?” 고민했었는데, 이번 기회에 정적 팩터리 메서드로 옮기는 쪽으로 정리하겠습니다.

1. WishResponseDto 클래스에 정적 팩터리 메서드 of(...) 추가

```java
// 변경 후 WishResponseDto.java
@Getter
@AllArgsConstructor(access = AccessLevel.PRIVATE)   // 외부에서 new 로 직접 만들지 못하도록 제한
public class WishResponseDto {

    private final Long productId;
    private final String productName;
    private final int price;
    private final String imageUrl;
    private final int quantity;

    /** WishItem → WishResponseDto 변환을 담당하는 정적 팩터리 메서드 */
    public static WishResponseDto of(WishItem wi) {
        return new WishResponseDto(
                wi.getProductId(),
                wi.getProductName(),
                wi.getPrice(),
                wi.getImageUrl(),
                wi.getQuantity()
        );
    }
}
```

이렇게 변환 로직을 DTO 내부에 두면 Service 레이어가 단순해지고, 변환 방식을 변경할 일이 생겨도 한 곳만 수정하면 됩니다.

2. WishService 에서 쓰던 변환 메서드를 삭제하고, 새 팩터리 메서드를 호출

```java
// 변경 전 WishService.java 일부
private WishResponseDto mapWishItemToWishResponseDto(WishItem wi) {
    return new WishResponseDto(
            wi.getProductId(),
            wi.getProductName(),
            wi.getPrice(),
            wi.getImageUrl(),
            wi.getQuantity()
    );
}

// 호출부
public  List<WishResponseDto>getWishListForMember(Member member) {
        List<WishItem> items = wishRepository.findAllByMemberId(member.getId());
        return items.stream()
                .map(this::mapWishItemToWishResponseDto)
                .toList();
    }
```

```java
// 변경 후 WishService.java 일부
// mapWishItemToWishResponseDto(...) 메서드 **삭제**

// 호출부
public  List<WishResponseDto>getWishListForMember(Member member) {
        List<WishItem> items = wishRepository.findAllByMemberId(member.getId());
        return items.stream()
                .map(WishResponseDto::of)
                .toList();
    }
```

변경 사항 반영해 커밋 올리겠습니다.

hjjae2 commented on src/main/java/gift/util/BearerAuthHeaderParser.java:
```diff
@@ -0,0 +1,25 @@
+        }
+
+        return authorizationHeader.substring(7).trim();
+
+    }
+
```
불필요한 공백은 제거하면 좋겠어요-!

5seonjae replied:
넵, 확인했습니다!
해당 파일 (  BearerAuthHeaderParser.java  ) 마지막 부분의 불필요한 공백 라인들은 제거하도록 하겠습니다.

```java
@Component
public class BearerAuthHeaderParser {

    public String extractBearerToken(String authorizationHeader) {
        if (authorizationHeader == null) {
            throw new MissingAuthorizationHeaderException("Authorization 헤더가 필요합니다.");
        }
        if (!authorizationHeader.startsWith("Bearer ")) {
            throw new InvalidAuthorizationHeaderException("Authorization 헤더 형식이 올바르지 않습니다.");
        }

        return authorizationHeader.substring(7).trim();
    }
}
```

정리 후 바로 커밋 올리겠습니다. 감사합니다! 😊

hjjae2 commented on src/main/java/gift/service/WishService.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    public void addWishItemForMember(Member member, WishRequestDto wishRequestDto) {
+        if (productRepository.findById(wishRequestDto.productId()).isEmpty()) {
+            throw new NoSuchElementException("상품을 찾을 수 없습니다.");
+        }
```
`member` 에 대해선 검증하지 않아도 되나요?

5seonjae replied:
### member 검증이 빠진 이유

컨트롤러에서 넘겨주는 member는 LoginMemberArgumentResolver 단계에서 이미 DB 조회로 검증합니다.

```java
@Component
public class LoginMemberArgumentResolver
    implements HandlerMethodArgumentResolver {

    private final TokenService tokenService;
    private final MemberRepository memberRepository;
    private final BearerAuthHeaderParser authHeaderParser;

    public LoginMemberArgumentResolver(TokenService tokenService,
                                       MemberRepository memberRepository,
                                       BearerAuthHeaderParser authHeaderParser) {
        this.tokenService = tokenService;
        this.memberRepository = memberRepository;
        this.authHeaderParser = authHeaderParser;
    }

    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return parameter.hasParameterAnnotation(LoginMember.class)
                && parameter.getParameterType().equals(Member.class);
    }

    @Override
    public Object resolveArgument(MethodParameter parameter,
                                  ModelAndViewContainer mavContainer,
                                  NativeWebRequest webRequest,
                                  WebDataBinderFactory binderFactory) {

        String header = webRequest.getHeader("Authorization");

        String token = authHeaderParser.extractBearerToken(header);

        Claims claims = tokenService.parseClaims(token);

        Long memberId = Long.valueOf(claims.getSubject());

        return memberRepository.findById(memberId)
                .orElseThrow(() -> new IllegalStateException("존재하지 않는 회원입니다."));
    }
}
```

1. JWT 파싱 → memberId 획득
2. `memberRepository.findById(memberId)` 
    - 없으면 즉시 MemberNotFoundException 발생 → Service 레이어로 내려오지 못함

따라서 WishService 시점에는 유효한 Member 엔티티가 보장되므로 중복 검증을 생략했습니다.

현 요구사항 범위에서는 Resolver → Member 검증, Service → Product 검증으로 책임을 분리해 두는 쪽을 우선 선택했습니다.
추가 방어선이 필요하다고 판단되면 말씀 주세요. 바로 반영하겠습니다! 😊

hjjae2 replied:
이 클래스/메서드 입장에서, `LoginMemberArgumentResolver` 로 검증된 `Member` 만 인자로 넘어올 수 있음을 보장할 수 있을까요-?

5seonjae replied:
이 메서드가 받는 Member는 컨트롤러 단계에서 @LoginMember 리졸버를 거치며 이미 유효성 검증(존재·상태)까지 완료된 엔티티입니다. 
웹 요청 경로로 진입하는 한 추가 검증은 중복이라고 판단했지만, 배치·테스트 등 다른 호출 경로가 생기면 래퍼 타입 도입이나 서비스 내부 재조회 등으로 방어층을 더 두겠습니다.

컨트롤러 단계에서 1차 검증이 끝나더라도,
**다른 호출 경로**에서 WishService가 직접 호출될 위험을 막기 위해
공통 validateMember 메서드를 두고 모든 Service 메서드에서 한 번 더 호출하도록 반영하겠습니다.
추후 배치·메시지 처리 모듈이 추가돼도 일관된 안전성을 유지할 수 있을 것으로 판단합니다.
추가 의견이나 더 강한 보장이 필요하면 알려 주세요! 😊

hjjae2 replied:
넵, 좋습니다. 👍 

애플리케이션 외부에서 들어오는 값을 신뢰하지 않듯 하나의 클래스, 메서드는 외부에서 들어오는 값(= 인자)을 신뢰하지 않는 것이 더 안전한 코드라고 의견을 드리고 싶습니다. 어떤 클래스의 앞/뒷단에서의 검증이나 제약사항은 언제든 변할 수 있는 요소이기 때문입니다. (다른 말로, 어떤 클래스나 제품은 앞/뒷단의 검증이나 제약사항에 의존하지 않고 독립적으로 안전한 클래스/제품이 되어야 합니다.)

선재님께서도 동의하신다면 앞으로도 이 점을 고려하시면서 작성해봐주셔도 좋을 것 같아요. 😄

hjjae2 commented on src/main/java/gift/repository/WishRepository.java:
WishRepository 에 대해서만이라도, 강의 시간에 언급된 `@JdbcTest`  를 사용해 테스트 코드를 작성해볼까요? 🙏

5seonjae replied:
알겠습니다! @JdbcTest 슬라이스 테스트로 WishRepository 만 검증하는 코드를 추가해 보겠습니다.

hjjae2 reviewed (changes requested):
안녕하세요, 선재님! 이번 단계도 깔끔하게 잘 구현해주셨습니다. 👍 

몇 가지 코멘트 남겼는데요, 확인 부탁드릴게요!

5seonjae committed:
style(auth): LoginMemberArgumentResolver 불필요한 주석 제거

- resolveArgument 메서드 내부의 설명 없는 주석 블록 삭제
- 코드 가독성 향상 및 불필요한 라인 제거

5seonjae committed:
refactor(wish): DELETE URI를 /api/wishes/{productId} → /api/wishes/product/{productId} 로 명확화

- **WishController**
    - @DeleteMapping 경로를 `{productId}` → `product/{productId}` 로 수정해 리소스 계층 구조를 명확히 표현
- **WishControllerTest**
    - 경로 변경에 맞춰 정상·예외(404/401) 시나리오 4종 모두 URL 업데이트

`/api/wishes/{id}` 형태만으로는 id가 **“위시 아이템”**인지 **“상품”**인지 혼동될 수 있어
`product/{productId}` 세그먼트를 삽입하여 의도를 명확히 했습니다.

5seonjae committed:
refactor(wish): WishItem 필드 명확화 — name → productName

- gift.entity.WishItem
    - 필드 `name`을 `productName`으로 리네임해 의미 명확화
    - 생성자·getter(`getName()` → 내부적으로 productName 반환) 코드 반영

5seonjae committed:
refactor(wish): WishResponseDto에 정적 팩토리 메서드 `of(WishItem)` 추가

- **WishResponseDto**
    - `static WishResponseDto of(WishItem wi)` 팩토리 메서드 구현
        → WishItem → DTO 변환 책임을 DTO 내부로 이동
- **WishService**
    - `mapWishItemToWishResponseDto()` 메서드 제거
    - `getWishListForMember()`에서 `items.stream().map(WishResponseDto::of)` 호출로 로직 간소화

- 변환 로직 위치가 명확해져 응집도 향상
- Service 레이어 코드 간결화 및 테스트 용이성 개선

5seonjae committed:
style: 전역 코드 포맷팅 및 불필요한 공백 정리

- IDE Code Reformat 기능으로 전체 소스 자동 정렬
- 들여쓰기/줄 끝 공백(trailing whitespace) 제거
- 중복·사용하지 않는 import 정리 및 개행 통일
- functional change 없음 (동작 로직 불변)

5seonjae committed:
refactor(wish): WishItem getter rename getName() → getProductName()

- gift.entity.WishItem
    - getter 메서드명을 getName() → getProductName() 으로 변경해 필드명과 일관성 확보
- WishResponseDto 및 관련 서비스·테스트 코드에서 호출부 전부 수정

5seonjae committed:
test(wish): WishRepository 단위 테스트 추가

- WishRepositoryTest (@JdbcTest + JdbcClient) 구현
    1. findAllByMemberId(): 초기 더미 데이터로 목록 조회 검증
    2. updateOrInsertWishItem(): 없는 상품 → INSERT, 이미 존재 시 수량 누적 UPDATE 확인
    3. delete(): memberId + productId 조건 삭제 시 영향 행 수 1, 이후 목록 비어 있음 검증
- H2 메모리 DB 사용, @BeforeEach 단계에서 members‧products‧wish_items 더미 데이터 삽입
- assertj-core 활용해 도메인 필드 값 및 행 수 단언

hjjae2 commented on src/main/java/gift/controller/api/WishController.java:
```diff
@@ -33,11 +33,10 @@ public void addWishItem(@LoginMember Member member,
     }
 
-    @DeleteMapping("/{productId}")
+    @DeleteMapping("/product/{productId}")
```
REST 디자인에 맞게 복수형으로 작성해볼까요?

hjjae2 commented on src/test/java/gift/WishRepositoryTest.java:
고생하셨습니다. 👍

hjjae2 reviewed (changes requested):
안녕하세요, 선재님! 코멘트 드린 내용에 대해 꼼꼼히 잘 확인해주셨네요. 👍 

마지막 단계인만큼, 남은 코멘트도 정리하고 마무리하면 좋겠는데요! 
확인 후 다시 코드 리뷰 요청 부탁드릴게요. 🙏

5seonjae committed:
refactor(wish): DELETE URI products/{productId}로 일관화 및 테스트 경로 수정

- WishController
    - @DeleteMapping 경로를 `product/{productId}` → `products/{productId}` 로 변경해
        POST/GET URI 구조와 복수형(products)으로 통일
- WishControllerTest
    - 네 가지 시나리오(정상·NotFound·Unauthorized·BadPrefix)에서
        요청 URL을 `/api/wishes/products/{productId}` 로 업데이트

5seonjae committed:
feat(wish): 회원 검증 로직 도입 및 InvalidMemberException 처리

- **InvalidMemberException** (gift.exception)
    - RuntimeException 기반 커스텀 예외 클래스 신설

- **GlobalExceptionHandler**
    - `@ExceptionHandler(InvalidMemberException.class)` 추가
        → 404 Not Found + `{ "error": "<message>" }` JSON 응답

- **WishService**
    - `validate(Member)` 메서드 추가 (member null → InvalidMemberException)
    - `getWishListForMember`, `addWishItemForMember`, `removeWishItemForMember`
        모두 호출 시점에 `validate()` 수행

- 컨트롤러 레이어 이전에 **회원 객체 검증**을 공통 처리
- 잘못된 인증/세션 누락 상황에서 일관된 404 응답 반환
- 서비스 메서드별 중복 null‑check 제거로 코드 간결화

5seonjae committed:
refactor(wish): WishService 메서드명 validate → validateMember 로 명확화

- 회원 null 검증 책임을 더 분명히 나타내도록 메서드명 변경
- getWishListForMember / addWishItemForMember / removeWishItemForMember 호출부 수정
- 기능 변화 없음 (동작 동일)

hjjae2 reviewed (approved):
선재님, 마지막 단계까지 고생 많으셨습니다.

미션을 진행하면서 보여주신 열정과 적극적인 의견 공유의 모습을 통해 저도 많이 배웠습니다. 👍 

남은 미션도 끝까지 화이팅입니다. 고생하셨습니다. 💪
